community/blog/2022-02 Property based testing the Arcus Python implementations for streamlined ML development/README.md

```diff
@@ -0,0 +1,150 @@
+![masked lung image](img/CHNCXR_0001_0_mask.png)
+
+The test itself is rather straightforward: load the images, transform them, verify the outcome. The images are represented here as an 3D-array (matrix of pixels) and ranges from 0 to 255 to have all the available RGB colors. Our function should transform this 3D-array to only zero (0) or one (1) values, representing either black or white. The `all_black_and_white` assertion function here will do that.
+
+```python
+import arcus.ml.images.io as ami
+import arcus.ml.images.conversion as conv
+
+image_path = 'tests/resources/images/lungs'
+
+def test_black_white_conversion_example():
+    image_list = ami.load_images(image_path)
+    image_converted = conv.to_blackwhite(image_list)
+    assert all_black_and_white(image_converted)
+```
+
+Before we can write our properties for our black/white function, we first need to generate our random images. Above, we saw that these images are nothing more than 3D [numpy](https://numpy.org/doc/stable/user/absolute_beginners.html) arrays. Fortunately, [Hypothesis has a way of expressing custom generators](https://hypothesis.readthedocs.io/en/latest/data.html#composite-strategies):
+
+```python
+import numpy as np
+from hypothesis import strategies as st
+
+@st.composite
+def images(draw, dimension = 3):
+    def create_array_rec(strategy, lengths, index = 0):
+        if index == len(lengths):
+            return strategy
+        length = lengths[index]
+        arr = st.lists(strategy, min_size=length, max_size=length).map(lambda x : np.array(x))
+        return create_array_rec(arr, lengths, index + 1)
+
+    num = st.integers(min_value=0, max_value=255)
+    amount = st.integers(min_value=1, max_value=10)
+    return draw(st.lists(amount, min_size=dimension, max_size=dimension).flatmap(lambda lengths : create_array_rec(num, lengths)))
+```
+
+Without going too deep into the actual implementation, this function will register for us a generator that can generate 3D [numpy](https://numpy.org/doc/stable/user/absolute_beginners.html) arrays where each cell contains a `float` value between 0 to 255. Exactly what our example images in the previous static example-based test represented. It's also worth mentioning that this implementation is a temporary result of a continuing process of improving the tests.
+
+That was the hardest part. Once you got your generator all set up, you can integrate it in your property test. Note that the default property here is even shorter than the initial static example-based test. The same assertion can also be reused here, proving a correct implementation of our black/white function.
+
+```python
+from hypothesis import given
+import arcus.ml.images.conversion as conv
+
+@given(images())
+def test_black_white_property(image_list):
+    image_converted = conv.to_blackwhite(image_list)
+    assert all_black_and_white(image_converted)
+```
+
+## Customizable time windows for more accurate predication
```

Comment:
predication or prediction ?

Reply:
prediction 😅 . Don't need to preach here. 😄